### PR TITLE
itunes input feature

### PR DESCRIPTION
**Pull request by brinda-wednesday**
### Ticket Link

---

### Related Links

---

### Description
Added an iTunes container with an input element used to search the song name
---

### Steps to Reproduce / Test

---

### GIF's

---
https://drive.google.com/file/d/1kL9JHeOxjRFaqTihJ7783X7y0ObSXdPz/view?usp=sharing

### CODE REVIEW

File: app/containers/HomeContainer/index.js [hunk 1]
```diff
@@ -55,6 +55,7 @@ export function HomeContainer({
 
   useEffect(() => {
     const loaded = get(reposData, 'items', null) || reposError;
+    console.log(loaded);
```
Comment (alichherawalla) [inline app/containers/HomeContainer/index.js]: remove console logs

File: app/containers/ITunes/index.js [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+import iTunesSaga from './saga';
+import { injectSaga } from 'redux-injectors';
+
+const { Search } = Input;
+const { Meta } = Card;
+
+export const SongsContainer = styled.div`
+  && {
+    margin: 20px 0;
+    max-width: 1000px;
```
Comment (alichherawalla) [inline app/containers/ITunes/index.js]: use em instead of px

File: app/containers/ITunes/index.js [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    flex-wrap: wrap;
+    justify-content: center;
+  }
+`;
+const ItunesCard = styled(Card)`
+  && {
+    margin: 10px 10px;
+    width: 400px;
+    border-radius: 20px;
+    background-color: #e5e7eb;
```
Comment (alichherawalla) [inline app/containers/ITunes/index.js]: const colors

File: app/containers/ITunes/index.js [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    position: relative;
+  }
+`;
+const Container = styled.div`
+  && {
+    max-width: 1000px;
```
Comment (alichherawalla) [inline app/containers/ITunes/index.js]: we're using max-width way too much. shouldn't width: 100% just suffice?

File: app/containers/ITunes/index.js [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    const items = get(songData, 'results', []);
+
+    return (
+      (items.length !== 0 || loading) && (
+        <SongsContainer>
+          {items.map((item, index) => (
+            <ItunesCard key={index}>
+              <Skeleton loading={loading} active>
+                <Meta
+                  avatar={<Avatar src={item.artworkUrl100} style={{ width: '75px', height: '75px' }} />}
```
Comment (alichherawalla) [inline app/containers/ITunes/index.js]: use styled-components. Don't use the style prop

File: app/containers/ITunes/index.js [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+                  title={item.trackName}
+                  description={item.artistName}
+                />
+
+                <audio controls style={{ width: '250px' }}>
```
Comment (alichherawalla) [inline app/containers/ITunes/index.js]: styled-components

File: app/containers/ITunes/index.js [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    width: 100%;
+    margin: 0 auto;
+    padding: 1.25em;
+  }
+`;
+
+export function ITunes({ songData = [], songError = null, songName, dispatchSongs, dispatchClearSongs }) {
+  const [loading, setLoading] = useState(false);
+  useEffect(() => {
+    const loaded = get(songData, 'results', null) || songError;
```
Comment (alichherawalla) [inline app/containers/ITunes/index.js]:
```
 const loaded = get(songData, 'results', songError);
```


if `loaded` is dependent on both `songData` as well as `songError` the dependency array should have both of them.

```
  }, [songData, songError]);
```

File: app/containers/ITunes/index.js [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+      dispatchSongs(sName);
+      setLoading(true);
+    } else {
+      dispatchClearSongs();
+    }
+  };
+  const debouncedHandleOnChange = debounce(handleOnChange, 200);
+
+  const renderSongList = () => {
+    const items = get(songData, 'results', []);
```
Comment (alichherawalla) [inline app/containers/ITunes/index.js]:
```
const items = get(songData, 'results', []);
return (
<If condition={items?.length || loading}>
...
...
</If>
```

File: app/containers/ITunes/selectors.js [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+/**
+ * Direct selector to the iTunes state domain
+ */
+
+const selectITunesDomain = (state) => state.iTunes || initialState;
+
+export const selectITunes = () => createSelector(selectITunesDomain, (substate) => substate);
+
+export const selectSongData = () => createSelector(selectITunesDomain, (substate) => get(substate, 'songData', null));
```
Comment (alichherawalla) [inline app/containers/ITunes/selectors.js]: don't use the third parameter for the lodash.get here. We don't need to supply it a default value as null. 

```
createSelector(selectITunesDomain, (substate) => get(substate, 'songData'));
```

please replicate for all selectors

File: app/containers/ITunes/tests/index.test.js [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+    await timeout(500);
+    expect(getiTunesSongsSpy).toBeCalled();
+    fireEvent.change(getByTestId('search-bar'), {
+      target: { value: '' }
+    });
+    await timeout(500);
+    expect(cleariTunesSongsSpy).toBeCalled();
+  });
+
+  it('should call dispatchSongs on change', async () => {
```
Comment (alichherawalla) [inline app/containers/ITunes/tests/index.test.js:43]: add more tests to ensure that if data is provided the grid is rendered correctly

Comment (alichherawalla) [reply app/containers/ITunes/tests/index.test.js:43]: number of children rendered in the grid are correct

File: app/utils/routeConstants.js [hunk 11]
```diff
@@ -6,5 +6,9 @@ export default {
       padding: 20
     },
     exact: true
+  },
+  itunes: {
+    route: '/itunes',
```
Comment (alichherawalla) [inline app/utils/routeConstants.js:11]: delete the package-lock.json. we use yarn

File: app/components/For/index.js [hunk 12]
```diff
@@ -35,6 +35,6 @@ For.propTypes = {
   noParent: Proptypes.bool
 };
 For.defaultProps = {
-  isRow: true
+  isrow: true
```
Comment (rameez-wed) [inline app/components/For/index.js]: Please use camelCase. I think isRow was correct

File: app/containers/ITunes/index.js [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+    if (!isEmpty(sName)) {
+      dispatchSongs(sName);
+      setLoading(true);
+    } else {
+      dispatchClearSongs();
+    }
+  };
+
+  const debouncedHandleOnChange = debounce(handleOnChange, 200);
+  const items = get(songData, 'results', []);
```
Comment (rameez-wed) [inline app/containers/ITunes/index.js]: Move these to the top

File: app/components/ItunesCard/index.js [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+    background-color: ${primarygrey};
+  }
+`;
+const CustomAvatar = styled(Avatar)`
+  && {
+    width: 5em;
+    height: 5em;
+  }
+`;
+export function ItunesCard({ songTitle, imgSrc, songArtist, audioSrc, trackId }) {
```
Comment (rameez-wed) [inline app/components/ItunesCard/index.js:41]: iTunesCard instead of ItunesCard

File: app/containers/ITunes/index.js [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+              trackId={item.trackId}
+            ></ItunesCard>
+          ))}
+        </GridLayout>
+      </If>
+    </Container>
+  );
+}
+
+ITunes.propTypes = {
```
Comment (rameez-wed) [inline app/containers/ITunes/index.js]: iTunes instead of ITunes

**Review by rameez-wed — CHANGES_REQUESTED** [review]

**Review by rameez-wed — APPROVED** [review]

**Pull request merged**